### PR TITLE
Feature: PoX Slot Events and Mempool Drop Tx Events

### PR DESCRIPTION
**Pull request by kantai**
This PR adds one new event type from dropped mempool txs (#2446), and adds a field to burn block events for the PoX reward slot holders (#2463).

### CODE REVIEW

File: docs/event-dispatcher.md [hunk 1]
```diff
@@ -100,10 +100,22 @@ Example:
...
   "burn_amount": 12000
 }
 ```
 
+* `reward_recipients` is an array of all the rewards received during this burn block. It may
+  include recipients who did _not_ have reward slots during the block. This could happen if
+  a miner's commitment was included a block or two later than intended. Such commitments would
+  not be valid, but the reward recipient would still receive the burn `amt`.
+* `reward_slot_holders` is an array of the Bitcoin addresses that would validly receive
+  PoX commitments during this block. These addresses may not actually receive rewards during
```
Comment (jcnelson) [inline docs/event-dispatcher.md:116]: Or, if miners simply don't mine in that block (e.g. fees are too high, etc.).

File: src/chainstate/stacks/miner.rs [hunk 2]
```diff
@@ -7369,7 +7388,8 @@ pub mod test {
 
                 let coinbase_tx = make_coinbase(miner, tenure_id as usize);
 
-                let mut anchored_block = StacksBlockBuilder::build_anchored_block(chainstate, &sortdb.index_conn(), &mut mempool, &parent_tip, tip.total_burn, vrf_proof, Hash160([tenure_id as u8; 20]), &coinbase_tx, ExecutionCost::max_value()).unwrap();
+                let mut anchored_block = StacksBlockBuilder::build_anchored_block(chainstate, &sortdb.index_conn(), &mut mempool, &parent_tip, tip.total_burn, vrf_proof, Hash160([tenure_id as u8; 20]), &coinbase_tx, ExecutionCost::max_value(), None
```
Comment (jcnelson) [inline src/chainstate/stacks/miner.rs:7391]: Wow, `cargo fmt` didn't split this into multiple lines?  Not a bug; just surprised at `cargo fmt`.

**Review by jcnelson — APPROVED** [review]
LGTM!  Thanks @kantai!  Should we require @zone117x sign off on this, so he can confirm the events are getting reported?

**Comment by hstove** [comment]
Does this PR also mean that, for blocks with no miner commits, `new_burn_block` will be emitted with `reward_slot_holders`?

**Comment by kantai** [comment]
> Does this PR also mean that, for blocks with no miner commits, new_burn_block will be emitted with reward_slot_holders?

Yep

File: docs/event-dispatcher.md [hunk 3]
```diff
@@ -100,10 +100,22 @@ Example:
...
   ],
+  "reward_slot_holders": [
+    "1C56LYirKa3PFXFsvhSESgDy2acEHVAEt6",
+    "1C56LYirKa3PFXFsvhSESgDy2acEHVAEt6"
+  ],
   "burn_amount": 12000
 }
 ```
 
+* `reward_recipients` is an array of all the rewards received during this burn block. It may
```
Comment (lgalabru) [inline docs/event-dispatcher.md:111]: I started working on similar in another branch a few weeks ago, and unless I missed something, if we want that list to be as exhaustive as possible, I think we'd need to tweak `LeaderBlockCommitOp::parse_from_tx` - ignoring a lot invalid block commit. I was thinking about producing some kind of `ConsensusReceipt` struct, that would help building a clear explanation of the consensus, and also returning the reasons for leader block commit being ignored.
Anyway this PR is way better than the approach I had in mind: it's live :).

**Review by lgalabru — APPROVED** [review]
LGTM!

File: testnet/stacks-node/src/event_dispatcher.rs [hunk 4]
```diff
@@ -204,6 +213,10 @@ impl EventObserver {
         self.send_payload(payload, PATH_MEMPOOL_TX_SUBMIT);
     }
 
+    fn send_dropped_mempool_txs(&self, payload: &serde_json::Value) {
+        self.send_payload(payload, PATH_MEMPOOL_TX_DROP);
```
Comment (zone117x) [inline testnet/stacks-node/src/event_dispatcher.rs:217]: Note this sends a lot of `StaleGarbageCollect` payloads without any transactions. Not a blocker just a tiny inefficiency.

**Review by zone117x — APPROVED** [review]
LGTM.
Tested dropped mempool with the API in https://github.com/blockstack/stacks-blockchain-api/pull/471 and reward slot holders in https://github.com/blockstack/stacks-blockchain-api/pull/473

Note I was only able to test dropped txs for replace-by-fee and stale-garbage-collect, the others I was unable to replicate in mainnet.

File: testnet/stacks-node/src/event_dispatcher.rs (hunk 4 above)
Comment (kantai) [reply testnet/stacks-node/src/event_dispatcher.rs:217]: Noted -- will add a check to make sure its only sending payloads when they're non-empty.

**Pull request merged**